controllers/topology/cluster_controller_test.go

```diff
@@ -0,0 +1,377 @@
+				WithVersion("1.22.2").
+				WithControlPlaneReplicas(3).
+				Build()).
+		Build()
+
+	// Create a set of initObjects from the objects above to add to the API server when the test environment starts.
+	initObjs := []client.Object{
+		clusterClass,
+		cluster,
+		infrastructureClusterTemplate,
+		infrastructureMachineTemplate,
+		bootstrapTemplate,
+		controlPlaneTemplate,
+	}
+	defer utilfeature.SetFeatureGateDuringTest(t, feature.Gates, feature.ClusterTopology, true)()
+
+	for _, obj := range initObjs {
+		err := env.Create(ctx, obj)
+		g.Expect(err).ToNot(HaveOccurred())
+	}
+	defer func() {
+		for _, obj := range initObjs {
+			err := env.Delete(ctx, obj)
+			g.Expect(err).ToNot(HaveOccurred())
+		}
+	}()
+	actualCluster := &clusterv1.Cluster{}
+	// Get Cluster object and check that relevant fields are set as expected.
+	g.Eventually(func() bool {
+		return assertClusterReconcile(ns.Name, clusterName, actualCluster)
+	}, timeout).Should(BeTrue())
+
+	// Check if InfrastructureCluster has been created and has the correct labels and annotations.
+	g.Eventually(func() bool {
+		return assertInfrastructureClusterReconcile(actualCluster)
+	}, timeout).Should(BeTrue())
+
+	// Check if ControlPlane has been created and has the correct version, replicas, labels and annotations.
+	g.Eventually(func() bool {
+		return assertControlPlaneReconcile(actualCluster, clusterClass)
+	}, timeout).Should(BeTrue())
+	g.Eventually(func() bool {
+		return assertMachineDeploymentsReconcile(actualCluster)
+	}, timeout).Should(BeTrue())
+}
+
+// assertClusterReconcile checks if the Cluster object:
+// 1) Has its InfrastructureReference and ControlPlane reference added correctly.
+// 2) InfrastructureReference and ControlPlaneRef have the expected Group, Version and Kind.
+func assertClusterReconcile(namespace, clusterName string, cluster *clusterv1.Cluster) bool {
+	key := client.ObjectKey{Name: clusterName, Namespace: namespace}
+	if err := env.Get(ctx, key, cluster); err != nil {
+		return false
+	}
+	// Check if relevant managed topology labels are present.
+	if ok := assertClusterTopologyOwnedLabel(cluster); !ok {
+		return false
+	}
+	// Check if InfrastructureRef exists and is of the expected Kind and APIVersion.
+	if ok := referenceExistsWithCorrectKindAndAPIVersion(cluster.Spec.InfrastructureRef,
+		builder.GenericInfrastructureClusterKind,
+		builder.InfrastructureGroupVersion); !ok {
+		log.Infof("We failed on infraref %v", cluster)
+		return false
+	}
+
+	// Check if ControlPlaneRef exists is of the expected Kind and APIVersion.
+	if ok := referenceExistsWithCorrectKindAndAPIVersion(cluster.Spec.ControlPlaneRef,
+		builder.GenericControlPlaneKind,
+		builder.ControlPlaneGroupVersion); !ok {
+		log.Infof("We failed on controlPlane ref %v", cluster)
+		return false
+	}
+	log.Infof("Cluster passed with %v", cluster)
+	return true
+}
+
+// assertInfrastructureClusterReconcile checks if the infrastructureCluster object:
+// 1) Is created correctly.
```

Comment:
Can/ should we expand what `correctly` means? (same for assertControlPlaneReconcile)

Reply:
Sure - I'll shorten to just 'is created' for both.